### ISSUE TITLE
importlib to replace imp module

### ISSUE THREAD

**Issue by AndyMender** [unknown-horizons/unknown-horizons#2667]
The Python imp module is being deprecated:
```py 
/usr/lib/python3.5/site-packages/nose/importer.py:12: PendingDeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
```

To that end, we can do the following:
A) remove the `imp` interface as it is used only for loading FIFE from a specific path defined via a command-line parameter (a branch with the respective adjustments is already there - remove-fife-commandline-loading)
B) seamlessly replace` imp` with` importlib` (new branch 2667-imp_to_importlib)

In an attempt to pursue B) I will:
- [x] read on the changes in `importlib`
- [x] implement `importlib` in a way that doesn't break our current setup (or use a slimmer workaround/alternative)

**Comment by squiddy** [comment]
That warning is about nose though, which will be removed soon anyway. Also, I've already replaced most uses of imp with importlib sometime in the last two weeks. There is a commit somewhere.

**Comment by squiddy** [comment]
You might be right with the Fife loading, I may not have checked that. :) Anyway, sounds good!

**Comment by AndyMender** [comment]
Yes, FIFE loading in run_uh.py (function `find_fife()`) still uses the legacy `imp` interface. The problem is that the Python Foundation has deprecated all of the simple interfaces to module finding and loading offered by `importlib` so now I need a workaround relying on classes and class methods like so: https://github.com/unknown-horizons/unknown-horizons/blob/d2c3ec475670f9b1dfd9fb19a7e22efd5a3f480e/run_tests.py#L46

I think combining the `importlib.machinery.PathFinder.find_spec()` method with `importlib.util.module_from_spec()` might be a solution.

**Comment by AndyMender** [comment]
I investigated the use of `importlib` a bit. The module is going through some rapid changes and basic functionalities are incomplete (`find_module` is deprecated in favor of `find_spec`, but it needs to be defined in a child class, because neither `PathFinder` nor `MetaPathFinder` provide it). I'll try to switch the interface to `importlib`, but will note in a comment that this may break soon anyway :/.

I am not very happy with the sudden explosion of boilerplate code...

**Comment by squiddy** [comment]
> EDIT: @squiddy, in such a case, should we keep the command-line interface for loading the FIFE module or should we assume FIFE will always be loaded from the default Python package directory?

It probably shouldn't be relevant for users. In the past IIRC some devs used the cmdline option.

> EDIT2: I discovered an interesting module called pluginbase (courtesy of tech2 from the python IRC channel), which seems to offer similar interfaces to importlib, but with lesser verbosity. I'll have a look whether this is an option as well. It may come in handy later when dealing with pyenet also!

Is this all really that complicated? All we're trying to do is import FIFE from a set of directories, right? Before using a third-party package, we might as well modify `sys.path` to include the directory and attempt an import.

```python
for directory in things_to_try:
    sys.path.insert(0, directory)
    # TODO cleanup sys.path after every try
    try:
       import fife
       break
     except ImportError:
       pass
else:
   print('Fife not found')
```

**Comment by AndyMender** [comment]
We might as well do that. I wanted to rely on modifying `sys.path` as a last resort, because that's usually discouraged :).

**Comment by squiddy** [comment]
True. But even though I haven't fully understood the whole importlib module, at **some** you have to point **something** (Loader, Finder whatever) at a directory to look into. I don't think those are magically looking everywhere in the file system.

**Comment by squiddy** [comment]
What I'm trying to say is, even though *explicit* `sys.path` manipulation might be discouraged, in this case it is probably justified. At least when compared to including a third-party package or 20 lines of `Finder` / `Spec` / `Loader` boilerplate. :)

Then again, importing python modules from some completely different non-standard directory good to begin with.

**Comment by AndyMender** [comment]
I will then implement a quick `.insert` and `.pop` for `sys.path` to transiently go through `path in paths` as you outlined :).

**Comment by squiddy** [comment]
> I was thinking of the pluginbase module as something we might need later for pyenet also when we decide to supply several modules bundled with Unknown Horizons. However, that we can leave for much later.

Yeah. I had a quick look at `pluginbase` and while it's definitely cool (I can imagine that's what powered the `from flask.ext import something` magic), right now it feels like overkill.

We'll see. I haven't tried the code I suggested, it might not work after all. :)

**Comment by AndyMender** [comment]
@squiddy, I added the corrections and will test today.

**Comment by AndyMender** [comment]
@squiddy, I added the corrections and will test today.

**EDIT:** Committed extra fixes and tested with the default FIFE loading directory at /usr/lib64/python3.6/site-packages/fife. I implemented a quick check to get just the parent directory surrounding the FIFE module as later we do `from fife import fife`.

**Comment by squiddy** [comment]
If you want feedback feel free to open a new pull request.

**Comment by AndyMender** [comment]
Implemented and merged: 498e49b8a5850f42a9b6dc07f83bdd306074993
